USERMANUAL.md

```diff
@@ -0,0 +1,599 @@
+    ALL \
+    configs \
+    --gcp_project_id="${GOOGLE_CLOUD_PROJECT}" \
+    --gcp_bq_dataset_id="${CLOUDDQ_BIGQUERY_DATASET}" \
+    --gcp_region_id="${CLOUDDQ_BIGQUERY_REGION}" \
+    --target_bigquery_summary_table="${CLOUDDQ_TARGET_BIGQUERY_TABLE}"
+```
+
+To see the resulting Data Quality validation summary statistics in the BigQuery table `$CLOUDDQ_TARGET_BIGQUERY_TABLE`, run:
+```bash
+#!/bin/bash
+export GOOGLE_CLOUD_PROJECT=$(gcloud config get-value project)
+export CLOUDDQ_BIGQUERY_REGION=EU
+export CLOUDDQ_TARGET_BIGQUERY_TABLE="<project-id>.<dataset-id>.<table-id>"
+echo "select * from \`${CLOUDDQ_TARGET_BIGQUERY_TABLE}\`" | bq query --location=${CLOUDDQ_BIGQUERY_REGION} --nouse_legacy_sql
+```
+
+To see the BigQuery SQL logic generated by CloudDQ for each `Rule Binding` stored as a view in `--gcp_bq_dataset_id`, run:
+```bash
+#!/bin/bash
+export CLOUDDQ_BIGQUERY_DATASET=clouddq
+bq show --view ${CLOUDDQ_BIGQUERY_DATASET}.T3_DQ_1_EMAIL_DUPLICATE
+```
+
+
+
+## Example Use Cases
+
+In this section, we will provide some example implementations of data quality checks that are common requirements.
+
+### Aggregate Conditions
+
+#### Unique Values Constraint
+
+Ensures that all values in a column are unique.
+
+```yaml
+rules:
+  UNIQUE:
+    rule_type: CUSTOM_SQL_STATEMENT
+    params:
+      custom_sql_statement: |-
+        select $column from data
+        group by $column
+        having count(*) > 1
+
+rule_bindings:
+  UNIQUENESS:
+    entity_id: TEST_DATA
+    column_id: UNIQUE_KEY
+    row_filter_id: NONE
+    rule_ids:
+      - UNIQUE
+```
+
+#### Unique Composed Value Constraint (over multiple columns)
+
+Ensures that the combination of values from several columns are all unique across the table.
+
+Note that the row count that is reported for this rule will be the number of elements that are duplicated. If one element is duplicated three times, the row count returned will be 1.
+
+```yaml
+rules:
+  UNIQUE_MULTIPLE:
+    rule_type: CUSTOM_SQL_STATEMENT
+    custom_sql_arguments:
+      - columns
+    params:
+      custom_sql_statement: |-
+        select $columns from data
+        group by $columns
+        having count(*) > 1
+        
+rule_bindings:
+  UNIQUENESS:
+    entity_id: TEST_DATA
+    row_filter_id: NONE
+    rule_ids:
+      - UNIQUE_MULTIPLE:
+          columns: unique_key, date
+```
+
+#### Maximum Gap
+
+Do not allow the gap between two subsequent values to be greater than a threshold.
+
+In this example, “two subsequent values” are two rows that are subsequent when sorting by the column. This works well for example when the column is a message timestamp, and we don’t want the time between messages to be too big.
+
+But if we have a value, for example, current, at different points in time, and we don’t want the current to jump more than a threshold value between two measurements, we need to change the order by clause in the custom_sql_statement, to order by the time column.
+
+```yaml
+rules:
+  GAP:
```

Comment:
prob not an ideal example as it hits exactly the limitation CloudDQ has at the moment - this is a row-level rule and all row-level rules should be expressed as CUSTOM_SQL_EXPRESSION currently, however the biggest (and probably the only) limitation of CUSTOM_SQL_EXPRESSION is its inability to accommodate logic that relies on context spanning across multiple records such as window functions (required for this rule)

Reply:
Discussed, actually this is a good set-based example and we should keep it.